internal/pkg/link/link.go

```diff
@@ -0,0 +1,151 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// -----------------------------------------------------------------------------
+
+package link
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/charmbracelet/lipgloss"
+	"github.com/charmbracelet/lipgloss/tree"
+	"github.com/protobom/protobom/pkg/sbom"
+	"github.com/protobom/storage/backends/ent"
+
+	"github.com/bomctl/bomctl/internal/pkg/db"
+	"github.com/bomctl/bomctl/internal/pkg/options"
+	"github.com/bomctl/bomctl/internal/pkg/sliceutil"
+)
+
+const cyan = lipgloss.ANSIColor(38)
+
+func AddLink(backend *db.Backend, opts *options.LinkOptions) error {
+	document, err := backend.GetDocumentByIDOrAlias(opts.ToIDs[0])
+	if err != nil {
+		return fmt.Errorf("%w", err)
+	}
+
+	if document == nil {
+		opts.Logger.Warn("Document not found", "id", opts.ToIDs[0])
+
+		return nil
+	}
+
+	switch {
+	case len(opts.DocumentIDs) > 0:
+		err = backend.AddDocumentAnnotations(opts.DocumentIDs[0], db.LinkToAnnotation, opts.ToIDs[0])
+	case len(opts.NodeIDs) > 0:
+		err = backend.AddNodeAnnotations(opts.NodeIDs[0], db.LinkToAnnotation, opts.ToIDs[0])
+	}
+
+	if err != nil {
+		return fmt.Errorf("%w", err)
+	}
+
+	return nil
+}
+
+func ClearLinks(backend *db.Backend, opts *options.LinkOptions) error {
+	for _, from := range opts.DocumentIDs {
+		if err := backend.RemoveDocumentAnnotations(from, db.LinkToAnnotation); err != nil {
+			return fmt.Errorf("%w", err)
+		}
+	}
+
+	for _, from := range opts.NodeIDs {
+		if err := backend.RemoveNodeAnnotations(from, db.LinkToAnnotation); err != nil {
+			return fmt.Errorf("%w", err)
+		}
+	}
+
+	return nil
+}
+
+func ListLinks(backend *db.Backend, opts *options.LinkOptions) error {
+	var (
+		annotations  ent.Annotations
+		id, fromType string
+		err          error
+	)
+
+	switch {
+	case len(opts.DocumentIDs) > 0:
+		fromType = "document"
+		id = opts.DocumentIDs[0]
+		annotations, err = backend.GetDocumentAnnotations(id, db.LinkToAnnotation)
+	case len(opts.NodeIDs) > 0:
+		fromType = "node"
+		id = opts.NodeIDs[0]
+		annotations, err = backend.GetNodeAnnotations(id, db.LinkToAnnotation)
+	}
+
+	if err != nil {
+		return fmt.Errorf("%w", err)
+	}
+
+	if len(annotations) == 0 {
+		opts.Logger.Warn("No links found for "+fromType, "id", id)
+
+		return nil
+	}
+
+	children := sliceutil.Extract(annotations, func(annotation *ent.Annotation) string {
+		return annotation.Value
+	})
+
+	style := lipgloss.NewStyle().Foreground(cyan)
+	links := tree.Root(fmt.Sprintf("Links for %s %s:", fromType, style.Render(id))).
+		Child(children).
+		EnumeratorStyle(style)
+
+	fmt.Fprintln(os.Stdout, links)
+
+	return nil
+}
+
+func RemoveLink(backend *db.Backend, opts *options.LinkOptions) error {
```

Comment:
Having some trouble getting this working: 

`link add` and `link list` worked fine, was able to add a link, then view it, but I can't seem to remove it: 
``` bash
$ go run . link remove document:urn:uuid:f360ad8b-dc41-4256-afed-337a04dff5db document:https://anchore.com/syft/file/bomctl_0.3.0
_linux_amd64.tar.gz-1b838d44-9d3c-47d0-9f7f-846397e701fa#DOCUMENT

$ go run . link list document:urn:uuid:f360ad8b-dc41-4256-afed-337a04dff5db
Links for document urn:uuid:f360ad8b-dc41-4256-afed-337a04dff5db:
└──https://anchore.com/syft/file/bomctl_0.3.0_linux_amd64.tar.gz-1b838d44-9d3c-47d0-9f7f-846397e701fa#DOCUMENT

$ go run . link remove document:https://anchore.com/syft/file/bomctl_0.3.0_linux_amd64.tar.gz-1b838d44-9d3c-47d0-9f7f-846397e701f
a#DOCUMENT document:urn:uuid:f360ad8b-dc41-4256-afed-337a04dff5db

$ go run . link list document:urn:uuid:f360ad8b-dc41-4256-afed-337a04dff5db
Links for document urn:uuid:f360ad8b-dc41-4256-afed-337a04dff5db:
└──https://anchore.com/syft/file/bomctl_0.3.0_linux_amd64.tar.gz-1b838d44-9d3c-47d0-9f7f-846397e701fa#DOCUMENT
```

As you can see I tried swapping the document ids, but still nothing. Might be missing something.

Reply:
Clear seems to work fine:

```bash 
$ go run . link clear document:urn:uuid:f360ad8b-dc41-4256-afed-337a04dff5db
$ go run . link list document:urn:uuid:f360ad8b-dc41-4256-afed-337a04dff5db
WARN  link: No links found for document id=urn:uuid:f360ad8b-dc41-4256-afed-337a04dff5db
```

Reply:
Good catch! Just pushed a fix, give that a shot

Reply:
Works! Thanks!